feat(msk): Cluster L2 Construct

curtiseppel opened the pull request:
L2 Construct for a MSK Cluster. 

I wrote this for internal use and thought I'd share it. I tried to follow the [example resource](https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/example-construct-library/lib/example-resource.ts) and [design guidelines](https://github.com/aws/aws-cdk/blob/master/DESIGN_GUIDELINES.md) as much as I could. Default properties were chosen either based on defaults when creating a cluster in the console or defaults set from CloudFormation.

Closes https://github.com/aws/aws-cdk/issues/9603

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ba75c8b45a517274cbc1874efce1ce1dd2c926f1
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0d8ba2ab-147b-4d6b-a5e1-095a02c2c190%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

CurtisEppel commented:
@iliapolo I've updated this PR to support the latest Kafka version and to support MSK's new authentication method - SASL/SCRAM. I also added a bunch of tests and the build is now passing.

iliapolo commented:
@CurtisEppel Thats awesome, thanks! Apologies the review is little lagging, its still on my radar though.

CurtisEppel commented on packages/@aws-cdk/aws-msk/lib/cluster-props.ts:
```diff
@@ -0,0 +1,340 @@
+ *
+ * @see https://docs.aws.amazon.com/msk/latest/developerguide/monitoring.html#metrics-details
+ */
+export enum ClusterMonitoringLevel {
```
I'm not sure if an enum is a good fit for this. This is the second time I've had to update this enum as a new monitoring level has been released (https://aws.amazon.com/about-aws/whats-new/2020/11/amazon-msk-offers-consumer-lag-metrics-select-topic-level-metrics-free/). Maybe should replace it with a class like [KafkaVersion](https://github.com/aws/aws-cdk/blob/ba75c8b45a517274cbc1874efce1ce1dd2c926f1/packages/%40aws-cdk/aws-msk/lib/cluster-version.ts) so that users can escape the enum with a custom value.

iliapolo reviewed (dismissed):
@CurtisEppel  I had several comments asking to flatten out configuration properties, in contrast to the CloudFormation structure. I just want to clarify the reasoning here.

Excessive nesting makes it hard to read and reason about, I know it allows for future proofing, but we normally don't do that. From prior experience, it's usually better to provide the best experience possible and employ deprecations in the future if necessary, than to sacrifice the ergonomics of the current API for some future unknown capabilities. 

Thanks for this PR!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 5fe6196475dfdee4d318b739e214c22d8a606dbf
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a15ec659-0b12-4e58-a98d-a166f20cf198%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f98484a6b3f139368ad20e233885dbfc60d72813
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e9a8fae8-839c-4d4d-8b52-b6041da35ea1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

CurtisEppel commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -0,0 +1,423 @@
+  /** Manages connections for the cluster */
+  public get connections(): ec2.Connections {
+    if (!this._connections) {
+      throw new Error('An imported Cluster cannot manage its security groups');
```
This was taken directly from the example - https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/example-construct-library/lib/example-resource.ts#L165

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 924ca39f10e80f1887a0f121f8123e1e6c37481a
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=91c7e863-5290-4693-a3b0-4f7403d01e06%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 3a029b4ede32a816b9048be6f0dd5ae59ad9e341
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=187e04e8-8868-4d52-9c94-b7cccd9a5550%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

CurtisEppel commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -0,0 +1,423 @@
+        dataVolumeKmsKeyId:
+            brokerNodeGroupProps.storageInfo.ebsStorageInfo.kmsKey.keyId,
+      }
+      : undefined; // MSK will create the managed key
```
> Amazon MSK always encrypts your data at rest.
https://docs.aws.amazon.com/msk/latest/developerguide/msk-encryption.html#msk-encryption-at-rest

CurtisEppel commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -0,0 +1,423 @@
+      const key = new kms.Key(this, 'SASLKey', {
+        description:
+          'Used for encrypting MSK secrets for SASL/SCRAM authentication.',
+        alias: 'msk/sasl/scram',
```
An alias would be nice, how about including the cluster name in the alias?

Side note the alias is actually not that useful as the MSK API doesn't allow you to associate secrets with the cluster that were created using the alias, even though Secrets Manager allows them to be created that way. I've created a support ticket and it has been verified as a known bug.

CurtisEppel replied:
I've included the cluster name

CurtisEppel commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -0,0 +1,423 @@
+            'kms:CreateGrant',
+            'kms:DescribeKey',
+          ],
+          resources: ['*'],
```
This policy gives users permission to use the key for the actions listed only when Secrets Manager makes the request on the user's behalf and doesn't allow any user to use the key directly. This means the resources can be `*` and the resources will instead be scoped by the users role if they have permission to read/write secrets.

CurtisEppel commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -0,0 +1,423 @@
+  private _zookeeperConnectionString(responseField: string): string {
+    const zookeeperConnect = new cr.AwsCustomResource(
+      this,
+      'ZookeeperConnect',
```
Good to know about lazy instances. Thanks :)

CurtisEppel commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -0,0 +1,423 @@
+        alias: 'msk/sasl/scram',
+      });
+
+      key.addToResourcePolicy(
```
https://docs.aws.amazon.com/kms/latest/developerguide/services-secrets-manager.html#asm-policies

CurtisEppel commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -0,0 +1,423 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
```
Can you clarify what should be moved to `cluster.ts`?

iliapolo replied:
Whoops. I meant to put this comment on the `cluster-props.ts` file. The properties of the cluster should be in the same file as the cluster.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 1d7b1bae3040e7f69d3186b71e496a513601ed71
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5bd52474-1d30-4722-b462-d30437e71f5d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 1e60eec6593a0091a9d7401023a681d8478af7f7
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=62f28e9e-f595-43d2-8a68-c353cfa80061%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8ebf0166a64130dd3e3778d7805bb52fe94d0931
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=480d559f-14cf-4f8f-8eb7-7094928e61e3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 0404bc5b85fc5482d77c4fbea664d4ad5ca2120d
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=87e9b05c-bdb8-4929-8768-84f63a18c6d5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

CurtisEppel commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -0,0 +1,423 @@
+        props.clientAuthenticationConfiguration?.sasl?.scram)
+    ) {
+      core.Annotations.of(this).addError(
+        'To enable client authentication, you must enabled TLS-encrypted traffic between clients and brokers.',
```
Nope you're right - it's simpler to just check if PLAINTEXT and clientAuthenticationConfiguration is provided

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 8bf2ae5a8e027e45661af2758fa2d11d37e813a7
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bbc3df0c-f6fd-40f2-ad5c-892bd5ba3d8e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

bracki commented:
Hey @iliapolo, @CurtisEppel, I have worked on MSK/Kafka event sources for Lambda here: https://github.com/aws/aws-cdk/issues/12099

Maybe we could have quick chat on the CDK slack or similar because I'm looking for feedback and because it would make sense if the EventSource can work with your implementation.
I could also provide code on how to setting up user/password-authentication with a AwsCustomResource for this PR. I'm "@bracki" on Slack. Thanks!

iliapolo commented:
@bracki Lets just have the discussion here, as i suspect it will be highly asynchronous and take a few days, this is a better place to keep track of it.

You are welcome to start it :)

bracki commented:
OK, my PR is ready and green. But it probably makes sense to wait for this PR to get ready. Currently I depend on `clusterArn:  string` but it makes more sense to take a Kafka cluster as an argument.

What does it take to get this merged?

iliapolo commented:
@bracki Yes I agree it makes sense to wait and use the L2 instead. This PR is under active review and I think should be merged fairly soon.

iliapolo commented:
@CurtisEppel Let me know when its ready for a second round? You can just `re-request` a review.

CurtisEppel commented:
> @CurtisEppel Let me know when its ready for a second round? You can just `re-request` a review.



> Hey @iliapolo, @CurtisEppel, I have worked on MSK/Kafka event sources for Lambda here: #12099
> 
> Maybe we could have quick chat on the CDK slack or similar because I'm looking for feedback and because it would make sense if the EventSource can work with your implementation.
> I'm "@bracki" on Slack. Thanks!

Nice, I also had an implementation of this but was holding off to add it to a separate PR once this one was through. I've been busy but will try and get this PR finished this week so that we can integrate your event source with this L2 construct.

CurtisEppel commented on packages/@aws-cdk/aws-msk/lib/cluster-props.ts:
```diff
@@ -0,0 +1,340 @@
+  /**
+   * Properties to be used for brokers in the cluster.
+   */
+  readonly brokerNodeGroupProps: BrokerNodeGroupProps;
```
Mainly following how CloudFormation grouped their properties. I'll see what flattening would look like.

> Are there any other nodes expect for brokers?
Not that I know of

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 722b792dd8ae514dac9101344375e5b9ef59d903
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c684b90f-14ee-4878-a052-5fc1358500cf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d91556e57d09d6ccfd210c8bfdd7b159ea1cb3c5
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2d575ff3-1ddd-40da-a099-525818af30e6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 3eb6aa007f7b4b3922d69975391c07afc6f4852b
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=56b85bb8-4353-427c-a328-bea148eb91e8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 35da90ea964dea4cd8cce819b05e0e0b9e10b6fa
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=71127366-bf75-4c2f-b64e-32531dbdc94e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b92bb0f3d1399aef6cc437c305c05ea53abaea2a
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b0605d73-a568-4ce4-b764-bf602e6b0477%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo commented on packages/@aws-cdk/aws-msk/test/cluster.test.ts:
```diff
@@ -0,0 +1,469 @@
+    vpc = new ec2.Vpc(stack, 'Vpc');
+  });
+
+  describe('created with default properties', () => {
```
All tests under this `describe` can be rolled into one since they just assert on different parts of the template, without actually configuring anything. Just a single test with a big assertion on all the properties would be better.

iliapolo commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -0,0 +1,771 @@
+   * @returns - The connection string to use to connect to the Apache ZooKeeper cluster.
+   */
+  private _zookeeperConnectionString(responseField: string): string {
+    return core.Lazy.stringValue({
```
Thats actually not what I meant. This would create a token each time it is called - and at synth time it would still create two constructs with the same id. 

Make sure you add a test for this.

What I meant is this:

```ts
  private _zookeeperConnectionString(responseField: string): string {
    if (!this._clusterDescription) {
      this._clusterDescription = new cr.AwsCustomResource(this, 'ZookeeperConnect', {
        onUpdate: {
          service: 'Kafka',
          action: 'describeCluster',
          parameters: {
            ClusterArn: this.clusterArn,
          },
          physicalResourceId: cr.PhysicalResourceId.of(
            'ZooKeeperConnectionString',
          ),
        },
        policy: cr.AwsCustomResourcePolicy.fromSdkCalls({
          resources: [this.clusterArn],
        }),
      });
    }
    return this._clusterDescription.getResponseField(`ClusterInfo.${responseField}`);
  }

```

I think it worth adding a `CfnOutput` for these attributes in the integ test, this would make sure the cr is invoked correctly.

iliapolo reviewed (dismissed):
@CurtisEppel Notice there are also some `README` violations in the build. You need to run `yarn pkglint` to fix.

iliapolo commented:
dont close

iliapolo commented:
@CurtisEppel Just making sure - are you still looking at this PR?

CurtisEppel commented:
> @CurtisEppel Just making sure - are you still looking at this PR?

Sorry @iliapolo, I've been busy at work with non-kafka things. I'll try and finish this off on this weekend.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 13b102ce25736d3cf2f57457bf1639ca60bb51c2
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=885c6726-496a-4fda-b74e-4618700d8baa%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: ed78ea38f56cf667a0f892d6d15c4aa6002aba31
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5a3b957c-affd-4515-aa46-0ab7aeeaf69e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

CurtisEppel commented on packages/@aws-cdk/aws-msk/test/integ.cluster.ts:
```diff
@@ -0,0 +1,20 @@
+  vpc,
+});
+
+cluster.connections.allowToAnyIpv4(
```
I think I left this in from when i was testing the IConnectable locally. I've removed it.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 4411138138d0ac21a0c575943072a72882758b58
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5d5d54ac-eab7-48a2-b12e-575226635460%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: e6d734f09cada1b110ba98e722cf44e081e5cc8d
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b74b5cfc-0515-43e1-b0d2-cda74a794f82%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 04b75ebe904319efd3b388d842eefc361f456930
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6e31198d-fb5c-431b-8b10-19bb61c4cee7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo commented:
@CurtisEppel When you're ready for another round just re-request a review 👍

CurtisEppel commented:
> @CurtisEppel When you're ready for another round just re-request a review 👍

I have requested a review...

The latest build failed due to a Jest timeout after I merged the latest master. Is this something that needs to be retried or did I actually break something?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: a23fa8fa5f0bca38ede35d39028853d882d8b329
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=efb8e656-849b-4dc9-9e69-819ddcaaba87%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 822068124c143d4726c3a9f8c16f23bb176fd5de
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b9bfc04f-d38d-4f27-8c69-5decbe2087b6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -0,0 +1,423 @@
+  ): ICluster {
+    class Import extends ClusterBase {
+      public readonly clusterArn = clusterArn;
+      public readonly clusterName = clusterArn.split('/')[1]; // ['arn:partition:kafka:region:account-id', clusterName, clusterId]
```
Actually, we can simply do:

```ts
public readonly clusterName = cdk.Fn.select(1, cdk.Fn.split('/', clusterArn));
```

To handle both tokens and concrete values.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 239b2ec2608b40c15720e3424c54b769c2335dc1
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=94534976-43cc-4640-a569-3372695a9904%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -1,34 +1,776 @@
+      !/^[a-zA-Z0-9]+$/.test(props.clusterName) &&
+      props.clusterName.length > 64
+    ) {
+      core.Annotations.of(this).addError(
```
Just throw instead of the annotation. Error annotations are normally used for validating lookup properties, i.e ones that can change outside of the synthesis process, this cannot. 

There are a couple of these around here, make sure you replace all of them.

iliapolo commented on packages/@aws-cdk/aws-msk/lib/cluster.ts:
```diff
@@ -1,34 +1,776 @@
+   */
+  public addUser(...usernames: string[]): void {
+    if (this.saslScramAuthenticationKey) {
+      const MSK_SECRET_PREFIX = 'AmazonMSK_'; // Required
```
Maybe add the cluster name? otherwise this will conflict when the same user is added to two different clusters.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 340f9018bfe4602d3a729f4b81c6f4a3bb687d78
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8d7496ce-6609-44c8-8192-6432c4a061e8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 22bac5c0593240843cb826afdd2456774019d06e
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=87b935e1-01a3-4f27-94f5-c8a3aa0c2014%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo reviewed (approved):
@CurtisEppel 

Awesome, thanks for all the hard work and sticking with this PR for so long :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 81c5eac49f6309f4dd07f0cb83ffc7704c40e3c0
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c5908f8d-44eb-413d-a27d-9fc426d68011%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

iliapolo commented:
@CurtisEppel Looks like there are some build failures after rebasing from master. Have a look?

CurtisEppel commented:
> @CurtisEppel
> 
> Awesome, thanks for all the hard work and sticking with this PR for so long :)

Thanks for all of the useful feedback 🎉 

@iliapolo  The snapshot test was failing due to lambda assets changing in the custom resource package. I've updated it now.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 86718ea267bdbb66320e509071a8260246df92de
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=41a47d79-7a22-4539-b833-088471ed57c7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 69baf6e3ecff0d666c02894198f1be353433fbe9
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a79433a8-4b11-4279-ac99-d942d258a7c2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: ae9c3d1a9be7fae66859db6a4ede7680865015d1
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fed57418-1559-4865-af3f-2b7d46765603%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
